stream.go

```diff
@@ -183,31 +90,79 @@ func SubscribeWith(lastEventID string, client *http.Client, request *http.Reques
 func SubscribeWithRequestAndOptions(request *http.Request, options ...StreamOption) (*Stream, error) {
 	defaultClient := *http.DefaultClient
 
-	stream := &Stream{
-		c:      &defaultClient,
-		req:    request,
-		retry:  DefaultInitialRetry,
-		Events: make(chan Event),
-		Errors: make(chan error),
-		closer: make(chan struct{}),
+	configuredOptions := streamOptions{
+		httpClient:         &defaultClient,
+		initialRetry:       DefaultInitialRetry,
+		retryResetInterval: DefaultRetryResetInterval,
 	}
 
 	for _, o := range options {
-		if err := o.apply(stream); err != nil {
+		if err := o.apply(&configuredOptions); err != nil {
 			return nil, err
 		}
 	}
 
+	var backoff backoffStrategy
+	var jitter jitterStrategy
+	if configuredOptions.backoffMaxDelay > 0 {
+		backoff = newDefaultBackoff(configuredOptions.backoffMaxDelay)
+	}
+	if configuredOptions.jitterRatio > 0 {
+		jitter = newDefaultJitter(configuredOptions.jitterRatio, 0)
+	}
+	retryDelay := newRetryDelayStrategy(
+		configuredOptions.initialRetry,
+		configuredOptions.retryResetInterval,
+		backoff,
+		jitter,
+	)
+
+	stream := &Stream{
+		c:           configuredOptions.httpClient,
+		lastEventID: configuredOptions.lastEventID,
+		readTimeout: configuredOptions.readTimeout,
+		req:         request,
+		retryDelay:  retryDelay,
+		Events:      make(chan Event),
+		Errors:      make(chan error),
+		Logger:      configuredOptions.logger,
+		closer:      make(chan struct{}),
+	}
+
 	// override checkRedirect to include headers before go1.8
 	// we'd prefer to skip this because it is not thread-safe and breaks golang race condition checking
 	setCheckRedirect(stream.c)
 
-	r, err := stream.connect()
-	if err != nil {
-		return nil, err
+	var initialRetryTimeoutCh <-chan time.Time
+	var lastError error
+	if configuredOptions.initialRetryTimeout > 0 {
+		initialRetryTimeoutCh = time.After(configuredOptions.initialRetryTimeout)
+	}
+	for {
```

Comment:
This logic allows us to retry the initial connection, if we've been configured to do so. Previously, if the first connection attempt failed it would immediately return an error.